### ISSUE TITLE
Binding data to a Json property named "Item" triggers an exception

### ISSUE THREAD

**Issue by warpech** [Starcounter/Starcounter#4029]
Starcounter 2.3.0.5456

I have just noticed that when I have such JSON:

```json
{
    "Html": "/Content/viewmodels/CmsItemPage.html",
    "Item": {
        "Key": ""
    },
    "Message$": "",
    "SaveClick$": 0
}
```

And I try to set the data to the `Item` property:

```cs
CmsItemPage page = new CmsItemPage();
var content = DbHelper.FromID(DbHelper.Base64DecodeObjectID(key)) as WebContent;
page.Item.Data = content;
return page;
```

I get the following exception:

```
System.Exception: ScErrCreateDataBindingForJson (SCERR14006): Cannot create binding between typed json and dataobject. Unable to generate binding. Json property 'CmsItemPage.Item' (Content.CmsItemPage+CmsItemWebContentPage), binding 'Item'.
Version: 2.3.0.5456.
Help page: https://github.com/Starcounter/Starcounter/wiki/SCERR14006.
   at Starcounter.XSON.TemplateDelegateGenerator.GenerateBoundDelegates(TObject property, Json json) in C:\TeamCity\BuildAgent\work\sc-996\Level1\src\Starcounter.XSON\TemplateDelegateGenerator.cs:line 287
   at Starcounter.Templates.TObject.GenerateBoundGetterAndSetter(Json parent) in C:\TeamCity\BuildAgent\work\sc-996\Level1\src\Starcounter.XSON\Templates\TObject.cs:line 91
   at Starcounter.Templates.TValue.UseBinding(Json parent) in C:\TeamCity\BuildAgent\work\sc-996\Level1\src\Starcounter.XSON\Templates\Foundation\TValue.cs:line 241
   at Starcounter.Templates.TObject.BoundOrUnboundGet(Json parent) in C:\TeamCity\BuildAgent\work\sc-996\Level1\src\Starcounter.XSON\Templates\TObject.cs:line 162
   at Content.CmsItemPage.get_Item() in C:\www\Content\src\Content\obj\x64\Debug\ViewModels\ViewModels\CmsItemPage.json:line 3
   at Content.MainHandlers.<>c.<RegisterPartials>b__1_1(String key) in C:\www\Content\src\Content\Api\MainHandlers.cs:line 87
   at lambda_method(Closure , Request , IntPtr , IntPtr )
   at Starcounter.Rest.UserHandlerInfo.RunUserDelegate(Request req, IntPtr methodSpaceUriSpaceOnStack, IntPtr parametersInfoOnStack) in C:\TeamCity\BuildAgent\work\sc-996\Level1\src\Starcounter.Rest\UriHandlersManager.cs:line 221
   at Starcounter.Internal.Web.AppRestServer.RunDelegateAndProcessResponse(IntPtr methodSpaceUriSpaceOnStack, IntPtr parametersInfoOnStack, Request req) in C:\TeamCity\BuildAgent\work\sc-996\Level1\src\Starcounter.Apps.JsonPatch\AppRestServer.cs:line 98
   at Starcounter.Rest.UriManagedHandlersCodegen.RunUriMatcherAndCallHandler(String methodSpaceUriSpace, String methodSpaceUriSpaceLower, Request req, UInt16 portNumber, Response& resp) in C:\TeamCity\BuildAgent\work\sc-996\Level1\src\Starcounter.Rest\UriManagedHandlersCodegen.cs:line 1117
   at Starcounter.Self.DoSelfCall(UInt16 portNumber, String method, String relativeUri, Dictionary`2 headersDictionary, String body, Byte[] bodyBytes, HandlerOptions handlerOptions, Request req) in C:\TeamCity\BuildAgent\work\sc-996\Level1\src\Starcounter.Internal\Rest\Self.cs:line 397
   at Starcounter.Self.GET(String uri) in C:\TeamCity\BuildAgent\work\sc-996\Level1\src\Starcounter.Internal\Rest\Self.cs:line 74
   at Content.StandalonePage.RefreshCurrentPage() in C:\www\Content\src\Content\ViewModels\StandalonePage.json.cs:line 18
   at Content.MainHandlers.<>c__DisplayClass0_0.<Register>b__5() in C:\www\Content\src\Content\Api\MainHandlers.cs:line 50
   at Starcounter.Db.<>c__DisplayClass45_0`1.<Scope>b__0() in C:\TeamCity\BuildAgent\work\sc-996\Level1\src\Starcounter\Db.cs:line 411
   at Starcounter.Db.Scope(Action action, Boolean isReadOnly) in C:\TeamCity\BuildAgent\work\sc-996\Level1\src\Starcounter\Db.cs:line 386
   at Starcounter.Db.Scope[TResult](Func`1 func, Boolean isReadOnly) in C:\TeamCity\BuildAgent\work\sc-996\Level1\src\Starcounter\Db.cs:line 412
   at Content.MainHandlers.<Register>b__0_4(String key) in C:\www\Content\src\Content\Api\MainHandlers.cs:line 45
   at lambda_method(Closure , Request , IntPtr , IntPtr )
   at Starcounter.Rest.UserHandlerInfo.RunUserDelegate(Request req, IntPtr methodSpaceUriSpaceOnStack, IntPtr parametersInfoOnStack) in C:\TeamCity\BuildAgent\work\sc-996\Level1\src\Starcounter.Rest\UriHandlersManager.cs:line 221
   at Starcounter.Internal.Web.AppRestServer.RunDelegateAndProcessResponse(IntPtr methodSpaceUriSpaceOnStack, IntPtr parametersInfoOnStack, Request req) in C:\TeamCity\BuildAgent\work\sc-996\Level1\src\Starcounter.Apps.JsonPatch\AppRestServer.cs:line 98
   at Starcounter.Internal.AppsBootstrapper.ProcessExternalRequest(Request req) in C:\TeamCity\BuildAgent\work\sc-996\Level1\src\Starcounter.Apps.JsonPatch\AppsBootstrapper.cs:line 765
HResult=-2146233088
HelpLink=https://github.com/Starcounter/Starcounter/wiki/SCERR14006
---> System.ArgumentException: Incorrect number of arguments supplied for call to method 'System.Object get_Item(Int32)'
   at System.Linq.Expressions.Expression.ValidateArgumentCount(MethodBase method, ExpressionType nodeKind, Int32 count, ParameterInfo[] pis)
   at System.Linq.Expressions.Expression.ValidateArgumentTypes(MethodBase method, ExpressionType nodeKind, ReadOnlyCollection`1& arguments)
   at System.Linq.Expressions.Expression.Call(Expression instance, MethodInfo method, IEnumerable`1 arguments)
   at Starcounter.XSON.TemplateDelegateGenerator.CreatePropertyGetBinding[T](PropertyInfo propertyInfo, MethodInfo getMethod, Expression variable, Boolean convertType, Template origin) in C:\TeamCity\BuildAgent\work\sc-996\Level1\src\Starcounter.XSON\TemplateDelegateGenerator.cs:line 663
   at Starcounter.XSON.TemplateDelegateGenerator.CreateGetMemberBinding[T](MemberInfo member, Expression variable, Boolean convertType, Template origin) in C:\TeamCity\BuildAgent\work\sc-996\Level1\src\Starcounter.XSON\TemplateDelegateGenerator.cs:line 553
   at Starcounter.XSON.TemplateDelegateGenerator.CreateBinding[T](BindingInfo bInfo, Boolean createGetBinding, Expression instance, ParameterExpression value, Template origin) in C:\TeamCity\BuildAgent\work\sc-996\Level1\src\Starcounter.XSON\TemplateDelegateGenerator.cs:line 523
   at Starcounter.XSON.TemplateDelegateGenerator.GenerateBoundGetExpression[T](BindingInfo bInfo, Template origin) in C:\TeamCity\BuildAgent\work\sc-996\Level1\src\Starcounter.XSON\TemplateDelegateGenerator.cs:line 412
   at Starcounter.XSON.TemplateDelegateGenerator.GenerateBoundDelegates(TObject property, Json json) in C:\TeamCity\BuildAgent\work\sc-996\Level1\src\Starcounter.XSON\TemplateDelegateGenerator.cs:line 260
HResult=-2147024809

More information about this error may be available in the server error log.
```

Renaming the property `Item` to anything else solves the problem.

cc @chrhol

**Comment by chrhol** [comment]
Thanks for finding this, it's clearly a bug.

It looks like it clashes with the indexer in `Json` (`json[int]`) that I think underneath creates two methods, `get_Item(int)` and `set_Item(int)`.

 Either the code that generates the binding needs to be fixed to pick the correct method or we need to block using the name `Item` as a property. 

I will take a look.

**Comment by andwah** [comment]
maybe related https://github.com/Starcounter/Starcounter/issues/3143  ?

**Comment by chrhol** [comment]
> maybe related #3143 ?

Yes, similar at least. I think this one might be avoided though, if the code that generates bindings picks the correct method.

**Comment by warpech** [comment]
It must be a regression because there is an app that used this successfully in 2.1: https://github.com/StarcounterPrefabs/Content/blob/master/src/Content/ViewModels/CmsItemPage.json#L3

**Comment by chrhol** [comment]
Fixed now, and pushed to v2.3 and develop.

**Comment by chrhol** [comment]
I will consider this issue solved and close it. Please reopen it if you find something that still isn't fixed for this issue.
